torch/distributed/_sharded_tensor/api.py

```diff
@@ -746,3 +923,120 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+        tensor([3, 4]) # Rank 0
+        tensor([3, 4]) # Rank 1
+
+        >>> # All tensors below are of torch.cfloat type.
+        >>> # We have 2 process groups, 2 ranks.
+        >>> tensor = torch.tensor([1+1j, 2+2j], dtype=torch.cfloat) + 2 * rank * (1+1j)
+        >>> tensor = torch.stack([tensor, tensor])
+        >>> tensor
+        tensor([[1.+1.j, 2.+2.j], [1.+1.j, 2.+2.j]]) # Rank 0
+        tensor([[3.+3.j, 4.+4.j], [3.+3.j, 4.+4.j]]) # Rank 1
+        >>> partial_tensor = _PartialTensor(tensor)
+        >>> complete_tensor = partial_tensor.reshard(collect_spec)
+        >>> complete_tensor
+        tensor([4.+4.j, 6.+6.j]) # Rank 0
+        tensor([4.+4.j, 6.+6.j]) # Rank 1
+    """
+
+    def __init__(self, local_partial_shard, pg, aggr_op=distributed_c10d.ReduceOp.SUM):
+        self.local_partial_shard = local_partial_shard
+        self.pg = pg
+        self.aggr_op = aggr_op
+
+    def __post_init__(self):
+        if not isinstance(self.local_partial_shard, torch.Tensor):
+            raise ValueError("local_partial_shard needs to be a Tensor.")
+        if not isinstance(self.aggr_op, distributed_c10d.ReduceOp):
+            raise ValueError(
+                "aggr_op needs to be a member of distributed_c10d.ReduceOp."
+            )
+
+    def reshard(self, resharding_spec: ShardingSpec) -> ShardedTensor:
+        """
+        Consolidate all partial tensors across multiple ranks and covert to a sharded
+        tensor based on the resharding spec.
+
+        Args:
+            resharding_spec (:class:`torch.distributed._sharding_spec.ShardingSpec`):
+                The specification describing how we reshard the aggregated local result.
+
+        Returns:
+            A :class:`ShardedTensor` filled with local aggregated result.
+        """
+        sharding_dim = int(resharding_spec.dim)  # type: ignore[attr-defined]
+        world_size = dist.get_world_size(self.pg)
+        local_shards = torch.tensor_split(self.local_partial_shard, world_size)
+        local_result = reduce_scatter(
+            torch.empty_like(local_shards[0]), list(local_shards), op=self.aggr_op
+        )
+
+        rank = dist.get_rank(self.pg)
+        local_metadata = None
+        shard_tensor_size = list(local_result.size())
+        shard_tensor_size[sharding_dim] *= world_size
+        current_offsets = [0] * len(local_result.size())
+        shards = []
+        for idx, placement in enumerate(resharding_spec.placements):  # type: ignore[attr-defined]
+            if rank == placement.rank():  # type: ignore[union-attr]
+                local_metadata = ShardMetadata(
+                    shard_offsets=current_offsets,
+                    shard_sizes=list(local_result.size()),
+                    placement=placement,
+                )
+                shards.append(Shard(local_result, local_metadata))
+                break
+            current_offsets[sharding_dim] += local_result.size(sharding_dim)  # type: ignore[index]
```

Comment:
Does this work for uneven sharding across ranks? For example if this code is run on the last rank where `local_result.size()` is probably smaller than the rest on `sharding_dim`?

Reply:
I thought about these cases for a while. For now, we have two assumptions:
1. World size needs to divide the length of the dimension to be resharded.
2. Each tensor has the same size across all ranks. 

The reason for these assumptions is that Partial Result comes from row-wise sharding linear op. It's not a sharded tensor and we don't contain a sharding spec here for how it is sharded previously. Plus, it looks like `reduce_scatter` does not support uneven size (different ranks have different tensor sizes). So I guess for now we are good?

Reply:
Can't we just use `get_chunked_dim_size` and `get_split_size` to get the right size for each partition given uneven split? We know the size of the complete tensor and which dim it is being sharded on so it should be possible to figure out the size/offset on each rank based on that instead of assuming world_size needs to divide length of dimension.

> Plus, it looks like reduce_scatter does not support uneven size (different ranks have different tensor sizes).

I'm assuming this only applies to the same index position in the list of inputs. But different indices in the input list can have different sizes?